trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,319 @@
+
+    return ndp
+
+
+def has_multiedges(G):
+    '''Determines if graph G contains multiple edges between any pair of
+    nodes. Returns 0 if False, 1 if True, and -1 if the provided graph is not
+    a NetworkX Multigraph.
+    '''
+    if G.is_multigraph():
+        H = nx.DiGraph(G)
+        if sorted(H.edges()) == sorted(G.edges()):
+            multiedges = 0
+        else:
+            multiedges = 1
+    else:
+        multiedges = -1
+
+    return multiedges
+
+
+def find_maximum_flow(G, s, t):
+    '''Finds maximum flow between a source and target node in graph G.
+    Requires edge attribute 'capacity'. Any edge without 'capacity' attribute
+    will be given infinite capacity and find_maximum_flow will not work as
+    intended.
+    '''    
+
+    warning_message = ("Warning: Maximum flow does not support " +
+        "MultiDiGraphs, the graph form that Trailmap uses. Graph will be " +
+        "converted to DiGraph, which can lose information if you have " +
+        "multiple edges running between two nodes. \nChecking if graph G " +
+        "has an identical DiGraph representation...\n")
+
+    if G.is_directed():
+        multiedges = has_multiedges(G)
+        if multiedges == True:
+            print(warning_message)
+            print("G appears to have multiple edges between at least one " +
+            "pair of nodes. Information WILL be lost in the switch to " +
+            "DiGraph. Proceeding with caution, consider the results at " +
+            "your own risk\n")
+            H = nx.DiGraph(G)
+        elif multiedges == False:
+            print(warning_message)
+            print("G does not appear to have multiple edges between any " +
+            "pair of nodes. No information should be lost in the switch to " +
+            "DiGraph. Proceeding with flow calculation.\n")
+            H = nx.DiGraph(G)
+        else:
+            H=G
+    else:
+        print("Provided graph is not a MultiDiGraph or DiGraph and other \
+        errors may occur. Abort.\n")
+        return None, None, G
+
+    max_flow_path = nx.maximum_flow(H, s, t)
+    max_flow = nx.maximum_flow_value(H, s, t)
+    return max_flow_path, max_flow, H
+
+
+def find_pathway_flow(G, pathway):
+    '''returns the maximum permissible flow for a given pathway. Any edge
+    without 'capacity' attribute will be given infinite capacity.
+    '''
+    multiedges = has_multiedges(G)
+    if multiedges != True:
+        # graph either has no multiedges, or is not in MultiDiGraph format
+        H = nx.DiGraph(G)
+        edges = set(pairwise(pathway))
+        #H.remove_nodes_from([n for n in G if n not in pathway])
+        H_sg = H.edge_subgraph(edges).copy()
+
+        sources = get_sources(H_sg)
+        sinks = get_sinks(H_sg)
+
+        # check for capacity in edges (if no capacity, NetworkX will raise an
+        # exception) and if every edge has infinite flow, return inf
+        infinite_flow = True
```

Comment:
Might be able to streamline/simplify all this by using `path_flow` as your logic variable and whether it `isinf()` as your test.